### PR TITLE
Refactoring with pure bash

### PR DESCRIPTION
**Pull request by tboerger**
This PR shows the work of months to propery handle go packaging. It's already in use at home:tboerger:branches:devel:languages:go and should be submitted to devel:languages:go soon. To keep the history clean you should just squash the commits. /cc @monstermunchkin

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -1,13 +1,17 @@
+== update version 15.0.0 ==
+
+  * Total refactoring of the codebase, dropped the ruby dependency and now we
+    are also using fure bash and go tooling
```
Comment (MaximilianMeister) [inline CHANGELOG]: `s/fure/pure`

File: golang.prov [hunk 2]
```diff
@@ -1,44 +1,21 @@
...
-# read stdin for filelist rpm feeds us for a (sub) package
-filelist = []
-prefix = buildroot + contribdir + "/"
-ARGF.each do |l|
-	# if line has "*.a"
-	# buildroot + contribdir + golang.org/x/text/collate/colltab.a
-	filelist << l.gsub(prefix,'').strip! if ( l.index(".a\n") && ! l.gsub(prefix + importpath,'').index(/example|test/) )
-end
+# /home/abuild/rpmbuild/BUILDROOT/golang-github-codegangsta-cli-1.17.1+git20160828.599ee69-0.x86_64/usr/share/go/contrib/src/github.com/codegangsta/cli
+# /home/abuild/rpmbuild/BUILDROOT/golang-github-codegangsta-cli-1.17.1+git20160828.599ee69-0.x86_64/usr/share/go/contrib/src/github.com/codegangsta/cli/altsrc
```
Comment (MaximilianMeister) [inline golang.prov]: remainder of debugging or are the comments somehow necessary?

File: golang.prov [hunk 3]
```diff
@@ -1,44 +1,21 @@
...
 
-# filelist:
-# golang.org/x/text/collate/colltab.a
+while read INPUT; do
```
Comment (MaximilianMeister) [inline golang.prov]: I'd put a comment there to explain why we need to read from STDIN here

File: golang.sh [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  local arch=$(uname -m)
+
+  case "${arch}" in
+  "x86_64")
+    echo "amd64"
+    ;;
+  "i386"|"i486"|"i586"|"i686"|"pentium3"|"pentium4"|"athlon"|"geode")
+    echo "386"
+    ;;
+  "armv3l"|"armv4b"|"armv4l"|"armv4tl"|"armv5b"|"armv5l"|"armv5teb"|"armv5tel"|"armv5tejl"|"armv6l"|"armv6hl"|"armv7l"|"armv7hl")
```
Comment (monstermunchkin) [inline golang.sh]: This can be shortened to `"armv*"`.

File: golang.prov (hunk 3 above)
Comment (MaximilianMeister) [reply golang.prov]: also smells like endless loop in some cases? at least i had this during testing e.g. `golang-github-exercism-cli` where it gets stuck after `Finding  Requires: /usr/lib/rpm/golang.req`

File: golang.sh [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+set -e
+set -x
+
+store_import_path() {
+  echo ${1} >| /tmp/importpath.txt
+}
+
+check_import_path() {
+  if [[ "$(get_import_path)" == "" ]]; then
+    echo "Empty import path, please specify a valid one!"
```
Comment (monstermunchkin) [inline golang.sh]: This should be redirected to stderr.

File: golang.prov (hunk 2 above)
Comment (tboerger) [reply golang.prov]: That's a reminder to see what gets piped in

File: golang.prov (hunk 3 above)
Comment (tboerger) [reply golang.prov]: No endless loop, just takes sometimes a little bit longer

Comment (MaximilianMeister) [reply golang.prov]: right, it really took quite a while but it worked fine :+1:

File: golang.sh [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    local destination=${RPM_BUILD_ROOT}$(get_source_path)${file#$(get_build_path)/src/}
+
+    echo "Copying ${file} to ${destination}"
+    install -D -m0644 ${file} ${destination}
+  done
+}
+
+process_test() {
+  if [[ "${1}" == "" ]]; then
+    echo "Please specify a valid importpath, refernce: go help test"
```
Comment (monstermunchkin) [inline golang.sh]: This should be redirected to stderr.

File: golang.sh [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    process_test ${@:2}
+    ;;
+  "--filelist"|"filelist")
+    process_filelist ${@:2}
+    ;;
+  "--godoc"|"godoc")
+    process_godoc ${@:2}
+    ;;
+  *)
+    echo "Please specify a valid method: arch, prep, build, install, source, test, filelist, godoc"
```
Comment (monstermunchkin) [inline golang.sh]: This should be redirected to stderr.

File: golang.sh [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+  GOPATH=$(get_build_path):$(get_buildcontrib_path) GOBIN=$(get_gobin_path) go \
+    test ${extra_flags} -x ${modifier}
+}
+
+process_filelist() {
+  local file_list="file.lst"
+
+  if [[ -f ${file_list} ]]; then
+    rm -f ${file_list}
```
Comment (monstermunchkin) [inline golang.sh]: Unnecessary `if`-block since `rm -f` will not fail if the file is not present.

**Review by monstermunchkin — CHANGES_REQUESTED** [review]

**Comment by tboerger** [comment]
Do you want to merge this? Or should I close that?

**Comment by monstermunchkin** [comment]
We'll merge it.

**Review by monstermunchkin — APPROVED** [review]

**Review by jordimassaguerpla — APPROVED** [review]

**Comment by tboerger** [comment]
Yeah!

**Pull request merged**